[IsDeprecatedWeakRefSmartPointerException] make BackgroundFetch related classes RefCounted

geoffreygaren opened the pull request:
#### 669e25c22543c8c82efd7b449718f1c2c7aed0ac
<pre>
[IsDeprecatedWeakRefSmartPointerException] make BackgroundFetch related classes RefCounted
<a href="https://bugs.webkit.org/show_bug.cgi?id=280194">https://bugs.webkit.org/show_bug.cgi?id=280194</a>
<a href="https://rdar.apple.com/136516715">rdar://136516715</a>

Reviewed by Ryosuke Niwa.

static_assert told me to.

* Source/WebCore/workers/service/background-fetch/BackgroundFetch.h:
* Source/WebCore/workers/service/background-fetch/BackgroundFetchEngine.cpp:
(WebCore::BackgroundFetchEngine::startBackgroundFetch):
(WebCore::BackgroundFetchEngine::resumeBackgroundFetch):
* Source/WebCore/workers/service/background-fetch/BackgroundFetchRecordLoader.h:
* Source/WebCore/workers/service/server/SWServer.h:
(WebCore::SWServer::createBackgroundFetchRecordLoader):
* Source/WebCore/workers/service/server/SWServerDelegate.h:
* Source/WebKit/NetworkProcess/BackgroundFetchLoad.h:
* Source/WebKit/NetworkProcess/NetworkDataTask.h:
* Source/WebKit/NetworkProcess/NetworkLoad.h:
* Source/WebKit/NetworkProcess/NetworkSession.cpp:
(WebKit::NetworkSession::createBackgroundFetchRecordLoader):
* Source/WebKit/NetworkProcess/NetworkSession.h:
* Source/WebKit/NetworkProcess/PingLoad.h:
* Source/WebKit/NetworkProcess/cocoa/NetworkSessionCocoa.mm:
(WebKit::NetworkSessionCocoa::loadImageForDecoding):
* Source/WebKit/NetworkProcess/storage/BackgroundFetchStoreManager.h:
(WebKit::BackgroundFetchStoreManager::create):
* Source/WebKit/NetworkProcess/storage/OriginStorageManager.cpp:
(WebKit::OriginStorageManager::StorageBucket::backgroundFetchManager):

Canonical link: <a href="https://commits.webkit.org/284174@main">https://commits.webkit.org/284174@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c3e636353cf258087c67fad95e4af5ba10dc570d

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/68669 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/48061 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/21328 "Built successfully") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/72738 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/19814 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/70786 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/55857 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/19630 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/72738 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/13172 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/71736 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/43918 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/59290 "Passed tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/5/builds/72738 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/40584 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/16714 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/18171 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/62537 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/17061 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/74432 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/12640 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/62/builds/16314 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/62226 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 vision-sim~~](https://ews-build.webkit.org/#/builders/86/builds/12680 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/59369 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/62253 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/15236 "Built successfully and passed tests") | [⏳ 🧪 vision-wk2 ](https://ews-build.webkit.org/#/builders/visionOS-1-Simulator-WK2-Tests-EWS "Waiting in queue, processing has not started yet") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/119/builds/3824 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/43862 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/44936 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/46130 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/44678 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

RupinMittal commented on Source/WebCore/workers/service/background-fetch/BackgroundFetch.h:
```diff
@@ -58,15 +49,24 @@ class SharedBuffer;
 struct CacheQueryOptions;
 
-class BackgroundFetch : public CanMakeWeakPtr<BackgroundFetch> {
+class BackgroundFetch : public RefCounted<BackgroundFetch>, public CanMakeWeakPtr<BackgroundFetch> {
```
Is it possible to use `RefCountedAndCanMakeWeakPtr` here?

rniwa replied:
Why not use RefCountedAndCanMakeWeakPtr?

RupinMittal commented on Source/WebCore/workers/service/background-fetch/BackgroundFetch.h:
```diff
@@ -84,6 +84,8 @@ class BackgroundFetch : public CanMakeWeakPtr<BackgroundFetch> {
     class Record final : public BackgroundFetchRecordLoaderClient, public RefCounted<Record>, private Identified<BackgroundFetchRecordIdentifier> {
         WTF_MAKE_TZONE_ALLOCATED(Record);
     public:
+        DEFINE_VIRTUAL_REFCOUNTED;
```
I'm confused about what this and `DECLARE_VIRTUAL_REFCOUNTED` do--just for my curiosity, would you mind explaining? Thanks!

geoffreygaren replied:
Haha, Chris was confused too. He talked me into removing it.

It's a way to conform to the protocol used by RefPtr using virtual ref/deref instead of concreted ref/deref. Then your subclass can implement the concrete ref/deref. You need this because a concreted subclass might inherit from multiple abstract base classes, but it can have only one refcount.

RupinMittal replied:
Ah got it. Thank you for explaining!

RupinMittal commented on Source/WebCore/workers/service/background-fetch/BackgroundFetchEngine.h:
```diff
@@ -29,25 +29,19 @@
 class SWServer;
 
-class BackgroundFetchEngine : public CanMakeWeakPtr<BackgroundFetchEngine> {
+class BackgroundFetchEngine : public RefCounted<BackgroundFetchEngine>, public CanMakeWeakPtr<BackgroundFetchEngine> {
```
Would it be possible to use `RefCountedAndCanMakeWeakPtr<BackgroundFetchEngine>` here too?

RupinMittal commented on Source/WebCore/workers/service/background-fetch/BackgroundFetchEngine.cpp:
```diff
@@ -74,8 +74,8 @@ void BackgroundFetchEngine::startBackgroundFetch(SWServerRegistration& registrat
-    auto& fetch = *result.iterator->value;
-    fetch.doStore([server = m_server, fetch = WeakPtr { fetch }, callback = WTFMove(callback)](auto result) mutable {
+    auto fetch = result.iterator->value;
+    fetch->doStore([server = m_server, fetch = WeakPtr { fetch }, callback = WTFMove(callback)](auto result) mutable {
```
Since `BackgroundFetch::doStore()` is non-trivial, perhaps we should hold a Ref of fetch on the stack?

geoffreygaren replied:
result.iterator->value is a Ref, so I believe this auto means Ref. (Whereas the old auto& would be a reference to the Ref in the table, which is not sufficient since the table can mutate.)

RupinMittal reviewed (commented):
Hello! I'm not a reviewer but thought I'd comment on a few things I noticed.

rniwa commented on Source/WebKit/NetworkProcess/storage/BackgroundFetchStoreManager.h:
```diff
@@ -31,26 +31,21 @@
 namespace WebKit {
 
-class BackgroundFetchStoreManager : public CanMakeWeakPtr<BackgroundFetchStoreManager> {
+class BackgroundFetchStoreManager : public RefCounted<BackgroundFetchStoreManager>, public CanMakeWeakPtr<BackgroundFetchStoreManager> {
```
Ditto. Use RefCountedAndCanMakeWeakPtr?

geoffreygaren committed:
[IsDeprecatedWeakRefSmartPointerException] make BackgroundFetch related classes RefCounted
https://bugs.webkit.org/show_bug.cgi?id=280194
rdar://136516715

Reviewed by Ryosuke Niwa.

static_assert told me to.

* Source/WebCore/workers/service/background-fetch/BackgroundFetch.h:
* Source/WebCore/workers/service/background-fetch/BackgroundFetchEngine.cpp:
(WebCore::BackgroundFetchEngine::startBackgroundFetch):
(WebCore::BackgroundFetchEngine::resumeBackgroundFetch):
* Source/WebCore/workers/service/background-fetch/BackgroundFetchRecordLoader.h:
* Source/WebCore/workers/service/server/SWServer.h:
(WebCore::SWServer::createBackgroundFetchRecordLoader):
* Source/WebCore/workers/service/server/SWServerDelegate.h:
* Source/WebKit/NetworkProcess/BackgroundFetchLoad.h:
* Source/WebKit/NetworkProcess/NetworkDataTask.h:
* Source/WebKit/NetworkProcess/NetworkLoad.h:
* Source/WebKit/NetworkProcess/NetworkSession.cpp:
(WebKit::NetworkSession::createBackgroundFetchRecordLoader):
* Source/WebKit/NetworkProcess/NetworkSession.h:
* Source/WebKit/NetworkProcess/PingLoad.h:
* Source/WebKit/NetworkProcess/cocoa/NetworkSessionCocoa.mm:
(WebKit::NetworkSessionCocoa::loadImageForDecoding):
* Source/WebKit/NetworkProcess/storage/BackgroundFetchStoreManager.h:
(WebKit::BackgroundFetchStoreManager::create):
* Source/WebKit/NetworkProcess/storage/OriginStorageManager.cpp:
(WebKit::OriginStorageManager::StorageBucket::backgroundFetchManager):

Canonical link: https://commits.webkit.org/284174@main
